stdlib.sh

```diff
@@ -26,6 +26,68 @@ direnv_config_dir=${XDG_CONFIG_DIR:-$HOME/.config}/direnv
 # algorithm and so it won't be re-exported.
 export DIRENV_IN_ENVRC=1
 
+# Usage: strict_env [COMMAND...]
+#
+# Turns on shell execution strictness. This will force the .envrc
+# evaluation context to exit immediately if:
+#
+# - any command in a pipeline returns a non-zero exit status that is
+#   not otherwise handled as part of `if`, `while`, or `until` tests,
+#   return value negation (`!`), or part of a boolean (`&&` or `||`)
+#   chain.
+# - any variable that has not explicitly been set or declared (with
+#   either `declare` or `local`) is referenced.
+#
+# If followed by a command-line, the strictness applies for the duration
+# of the command.
+#
+# Example:
+#
+#    strict_env
+#    has curl
+#
+#    strict_env has curl
+strict_env() {
+  set -euo pipefail
+
+  if (( $# )); then
+    local res
+
+    "${@}"
+
+    res=$?
+
+    set +euo pipefail
```

Comment:
Do you know if it's possible to detect and backup the original state of `eu` and `pipefail` so they can be restored here?

Reply:
Wouldn't this help (`help set`)?
```
The current set of flags may be found in $-.
```

Reply:
`$-` doesn’t contain the `set -o pipefail` flag. That can be obtained (and parsed) via `set -o`.

Reply:
@zimbatm Yes, it is. If I were to do that, I’d probably want to use `set -o errexit -o nounset -o pipefail` to set, because that’s the corresponding values from `set -o`.

Reply:
OK. The latest version does this mostly correctly. I have tested this in an interactive bash shell (the output of `set -o` has been trimmed to the relevant settings):

```sh
bash-3.2$ source stdlib.sh
bash-3.2$ set -o
errexit        	off
nounset        	off
pipefail       	off

bash-3.2$ strict_env
bash-3.2$ set -o
errexit        	on
nounset        	on
pipefail       	on

bash-3.2$ strict_env echo HELLO
HELLO
bash-3.2$ set -o
errexit        	off
nounset        	on
pipefail       	on

bash-3.2$ unstrict_env echo HELLO
HELLO

bash-3.2$ set -o
errexit        	off
nounset        	on
pipefail       	on

bash-3.2$ unstrict_env
bash-3.2$ set -o
errexit        	off
nounset        	off
pipefail       	off
```

The one piece that’s _broken_ is `set -o errexit` and that‘s because the execution in a subshell doesn’t preserve `set -e`.